keps/sig-api-machinery/3285-subresource-finalizers/README.md

```diff
@@ -0,0 +1,784 @@
+  - `labels`
+  - `annotations`
+  - `ownerReferences`
+  - `status/conditions`
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Add `/{resource_name}/finalizers` endpoint to all resources and allow access to `metadata.finalizers` for the resources through the endpoints.
+The actual field in each resource isn't added, but the request to the endpoint is handled as the operation to `metadata.finalizers`.
+`Role` and `ClusterRole` can be set to the endpoint, therefore users can set a separate RBAC to `/{resource_name}/finalizers`.
+A PoC implementation can be found [here](https://github.com/mkimuram/kubernetes/commit/751a67e03d62397c1d21947e9e2e3a81f279a29f) (Please also see [here](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) for how it works. Note that API definition in this KEP will differ from the one in the PoC).
+
+### API Changes
+
+A new `Finalize` resource is introduced in `subresources` API group:
+
+```golang
+// Finalize represents a finalizers field to be accessed via subresource.
+type Finalize struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Only the `finalizers` field is updateable.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+```
+
+### Supported API calls to the endpoint
+
+The `/{resource_name}/finalizers` endpoint supports 3 verbs, GET, PATCH, and PUT.
+
+#### GET
+
+GET verb is used to read the value in `metadata.finalizers` field.
+
+- Request format:
+Empty string is passed as a request.
+
+- Response format:
+  - On success, the `Finalize` resource is returned.
+    ex)
+    ```json
+    {
+        "kind": "Finalize",
+        "apiVersion": "v1alpha1",
+        "metadata": {
+            "finalizers": ["example.com/my-finalizer"],
+            "creationTimestamp": null,
+		}
+    }
+    ```
+
+  - On failure, error message is stored in `Status` resource.
+    ex)
+    ```json
+    {
+        "kind": "Status",
+        "apiVersion": "v1",
+        "metadata": {},
+        "status": "Failure",
+        "message": "\"finalize\" is forbidden: User \"example-user\" cannot get subresource \"finalizers\" of \"example\" in API group \"example\" in the namespace \"example\"",
+        "reason": "Forbidden",
+        "details": {},
+        "code": 403
+    }
+    ```
+
+- Status code:
+  - 200 "OK": Returned on success without change
+  - 403 "Forbidden": Returned when GET verb is not allowed for `/{resource_name}/finalizers` endpoint (RBAC admission handles this)
+
+#### PATCH
```

Comment:
test-cases: server-side apply, client-side apply, and other patch variants.

Reply:
Added as e2e test cases.